util/publicdns/publicdns.go

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package publicdns contains mapping and helpers for working with
+// public DNS providers
+package publicdns
+
+import "inet.af/netaddr"
+
+var KnownDoH = map[netaddr.IP]string{} // 8.8.8.8 => "https://..."
+
+var DohIPsOfBase = map[string][]netaddr.IP{}
+
+func addDoH(ipStr, base string) {
+	ip := netaddr.MustParseIP(ipStr)
+	KnownDoH[ip] = base
+	DohIPsOfBase[base] = append(DohIPsOfBase[base], ip)
+}
+
+func DohV6(base string) (ip netaddr.IP, ok bool) {
+	for _, ip := range DohIPsOfBase[base] {
+		if ip.Is6() {
+			return ip, true
+		}
+	}
+	return ip, false
+}
+
+func init() {
```

Comment:
so in general, best practice is to avoid init funcs because they slow startup of programs that import this package and don't use it. It turns out our two importing packages will always use it, but we might want to get the API right to start, returning accessor functions that use a `sync.Once` to do this func once if/when needed. Then the linker can see when they're not used and discard it all.

so one package-level:

```go
var populateOnce sync.Once
```

And then name this `func init()` as `func populate()` and then at accessors, at the top:

```go
   populateOnce.Do(populate)
   return ...
```

Reply:
Ah good to know, thanks!